include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h

```diff
@@ -719,6 +724,264 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
+
+    auto event = __queue.submit([&](sycl::handler& hdl) {
+        auto wg_copy_if_values = sycl::local_accessor<_Type, 1>(sycl::range<1>{elems_in_tile}, hdl);
+
+        oneapi::dpl::__ranges::__require_access(hdl, __in_rng, __out_rng, __num_rng);
+        hdl.parallel_for(sycl::nd_range<1>(num_workitems, wgsize), [=](const sycl::nd_item<1>& item)  [[intel::reqd_sub_group_size(SUBGROUP_SIZE)]] {
+            auto group = item.get_group();
+            auto wg_local_id = item.get_local_id(0);
+            constexpr ::std::uint32_t stride = wgsize;                 
+                                                            
+            // Global load into local
+            auto wg_current_offset = 0;
+
+            _SizeT wg_count = 0;
+
+            // Phase 1: Create wg_count and construct in-order wg_copy_if_values
+            if (elems_in_tile <= n) {
+#pragma unroll
+              for (size_t i = 0; i < elems_in_tile; i += wgsize) {
+                _Type val = __in_rng[i + wg_local_id];
+
+                _SizeT satisfies_pred = pred(val);
+                _SizeT count = sycl::exclusive_scan_over_group(group, satisfies_pred, wg_count, sycl::plus<_SizeT>());
+
+                if (satisfies_pred)
+                  wg_copy_if_values[count] = val;
+
+                wg_count = sycl::group_broadcast(group, count + satisfies_pred, wgsize - 1);
+              }
+            } else {
+              // Edge of input, have to handle memory bounds
+              // Might have unneccessary group_barrier calls
+#pragma unroll
+              for (size_t i = 0; i < elems_in_tile; i += wgsize) {
+                _SizeT satisfies_pred = 0;
+                _Type val = *std::launder(reinterpret_cast<_Type*>(alloca(sizeof(_Type))));
+                if (i + wg_local_id < n) {
+                  val = __in_rng[i + wg_local_id];
+
+                  satisfies_pred = pred(val);
+                }
+                _SizeT count = sycl::exclusive_scan_over_group(group, satisfies_pred, wg_count, sycl::plus<_SizeT>());
+
+                if (satisfies_pred)
+                  wg_copy_if_values[count] = val;
+
+                wg_count = sycl::group_broadcast(group, count + satisfies_pred, wgsize - 1);
+              }
+            }
+
+            // Phase 3: copy values to global memory
+            for (int i = wg_local_id; i < wg_count; i += wgsize) {
+                __out_rng[i] = wg_copy_if_values[i];
+            }
+            if (group.leader())
+                __num_rng[0] = wg_count;
+        });
+    });
+
+    event.wait();
+}
+
+template <typename _KernelParam, typename _UseAtomic64, typename _UseDynamicTileID, typename _InRange, typename _OutRange, typename _NumSelectedRange, typename _UnaryPredicate>
+void
+single_pass_copy_if_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __out_rng, _NumSelectedRange __num_rng, _UnaryPredicate pred)
+{
+    using _Type = oneapi::dpl::__internal::__value_t<_InRange>;
+    using _SizeT = uint64_t;
```

Comment:
Don't worry about this for now, but could be a ticket: For cases up to 2^30, we can perform the lookback with a single 32-bit int (2 bits for flags). Could template and runtime dispatch depending on `n`.

Reply:
Made ticket